File: trunk/CodeModel/Language/AS3/AS3Project.cs
/*
Copyright (c) 2008 Joa Ebert

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

using System;
using System.Collections.Generic;
using System.Text;
using AS3V.CodeModel.Generic;
using AS3V.CodeModel.Exceptions;
using System.IO;
using AS3V.CodeModel.Generic.Enumerators;

namespace AS3V.CodeModel.Language.AS3
{
    public class AS3Project : IProject
    {
        private string _name;
        private string _path;

        private List<AS3PackageRoot> _roots;

        public AS3Project()
        {
            _name = "";
            _path = "";
            _roots = new List<AS3PackageRoot>();
        }

        public void Build(string path, List<string> sourceFolder)
        {
            if (path.LastIndexOf('/') != path.Length - 1)
            {
                path += '/';
            }

            if (!Directory.Exists(path))
            {
                throw new Exception(path + " does not exist.");
            }

            _name = new FileInfo(path.Substring(0,path.Length - 1)).Name;
            _path = path;

            int n = sourceFolder.Count;

            for (int i = 0; i < n; ++i)
            {
                string rootPath;

                if (sourceFolder[i][1] != ':')
                {
                    rootPath = path + sourceFolder[i];
                }
                else
                {
                    rootPath = path;
                }

                AS3PackageRoot root = new AS3PackageRoot();

                root.Build(this, null, rootPath);

                _roots.Add(root);                
            }
        }

        #region IProject Member

        public int RootCount
        {
            get { return _roots.Count; }
        }

        public IPackageRoot RootAt(int index)
        {
            if (index < 0 || index >= RootCount)
                throw new RangeException(index, RootCount);

            return _roots[index];
        }

        public IEnumerable<IPackageRoot> RootEnumerator
        {
            get
            {
                return new PackageRootEnumerator(this);
            }
        }

        public string Name
        {
            get
            {
                return _name;
            }
        }

        public string DirectoryPath
        {
            get
            {
                return _path;
            }
        }

        public bool HasCompilationUnit(string unitPath)
        {
            return null != CompilationUnit(unitPath);
        }

        public ICompilationUnit CompilationUnit(string unitPath)
        {
            int charIndex = unitPath.LastIndexOf('.');

            if (-1 == charIndex)
            {
                foreach (AS3PackageRoot root in RootEnumerator)
                {
                    ICompilationUnit unit = root.CompilationUnit(unitPath);

                    if (null != unit)
                    {
                        return unit;
                    }
                }

                return null;
            }
            else
            {
                string unitName = unitPath.Substring(charIndex + 1);
                string packagePath = unitPath.Substring(0, charIndex);

                IPackage package = Package(packagePath);

                return package.CompilationUnit(unitName);
            }
        }

        public bool HasPackage(string packagePath)
        {
            return null != Package(packagePath);
        }

        public IPackage Package(string packagePath)
        {
            IPackage package;
            string[] pathElements = packagePath.Split('.');

            int i;
            int n = pathElements.Length;

            foreach (AS3PackageRoot root in RootEnumerator)
            {
                package = root;

                for (i = 0; i < n; ++i)
                {
                    package = package.Package(pathElements[i]);

                    if (null == package)
                    {
                        break;
                    }
                }

                if (null != package)
                {
                    return package;
                }
            }

            return null;
        }

        public void Parse()
        {
            foreach(IPackageRoot root in RootEnumerator)
            {
                ParsePackage(root);
            }
        }

        private void ParsePackage(IPackage package)
        {
            foreach (ICompilationUnit compilationUnit in package.CompilationUnitEnumerator)
            {
                compilationUnit.Parse();
            }

            foreach (IPackage nextPackage in package.PackageEnumerator)
            {
                ParsePackage(nextPackage);
            }
        }

        #endregion
    }
}

File: trunk/java/src/com/joa_ebert/as3v/language/as3/AS3Project.java
package com.joa_ebert.as3v.language.as3;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;

import com.joa_ebert.as3v.projectModel.ICompilationUnit;
import com.joa_ebert.as3v.projectModel.IPackage;
import com.joa_ebert.as3v.projectModel.IPackageRoot;
import com.joa_ebert.as3v.projectModel.IProject;

public final class AS3Project implements IProject {
	private String _name = "";
	private String _path = "";
	private ArrayList<IPackageRoot> _roots = new ArrayList<IPackageRoot>();

	/**
	 * Creates a new AS3Project instance.
	 */
	public AS3Project() {
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public void build(String filePath, ArrayList<String> packageRoots)
			throws IOException {
		if (!filePath.endsWith(File.separator))
			filePath += File.separator;

		File[] systemRoots = File.listRoots();
		
		File project = new File(filePath);

		if (!project.exists()) {
			throw new IOException("Project path does not exist.");
		}
		
		_name = project.getName();
		_path = filePath;

		for(Iterator<String> iter = packageRoots.iterator(); iter.hasNext();)
		{
			String rootPath = iter.next();
			
			boolean isAbsolute = false;
			
			for(int i = 0; i < systemRoots.length; ++i)
			{
				if(rootPath.startsWith(systemRoots[i].getAbsolutePath()))
				{
					isAbsolute = true;
					break;
				}
			}

			if(!isAbsolute)
			{
				rootPath = filePath + rootPath;
			}
			
			AS3PackageRoot root = new AS3PackageRoot();
			
			root.build(this, null, rootPath);
			
			_roots.add(root);
		}
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public ArrayList<ICompilationUnit> compilationUnit(String unitPath) {
		ArrayList<ICompilationUnit> results = new ArrayList<ICompilationUnit>();
		final int charIndex = unitPath.lastIndexOf('.');

		if (-1 == charIndex) {
			for (Iterator<IPackageRoot> iter = rootIterator(); iter.hasNext();) {
				ICompilationUnit unit = iter.next().compilationUnit(unitPath);

				if (null != unit)
					results.add(unit);
			}
		} else {
			String unitName = unitPath.substring(charIndex + 1);
			String packagePath = unitPath.substring(0, charIndex);

			ArrayList<IPackage> packages = package$(packagePath);

			if (null != packages) {
				for (Iterator<IPackage> iter = packages.iterator(); iter
						.hasNext();) {
					IPackage package$ = iter.next();
					ICompilationUnit unit = package$.compilationUnit(unitName);
					if (null != unit) {
						results.add(unit);
					}
				}
			}
		}

		return results;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public String filePath() {
		return _path;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public boolean hasCompilationUnit(String unitPath) {
		return 0 != compilationUnit(unitPath).size();
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public boolean hasPackage(String packagePath) {
		return 0 != package$(packagePath).size();
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public String name() {
		return _name;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public ArrayList<IPackage> package$(String packagePath) {
		ArrayList<IPackage> results = new ArrayList<IPackage>();
		String[] pathElements = packagePath.split(".");

		int i;
		final int n = pathElements.length;

		for (Iterator<IPackageRoot> iter = rootIterator(); iter.hasNext();) {
			IPackage package$ = iter.next();

			for (i = 0; i < n; ++i) {
				package$ = package$.package$(pathElements[i]);

				if (null == package$)
					break;
			}

			if (null != package$) {
				results.add(package$);
			}
		}

		return results;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public IPackageRoot rootAt(int index) {
		if (index < 0 || index >= _roots.size())
			return null;
		return _roots.get(index);
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public int rootCount() {
		return _roots.size();
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Iterator<IPackageRoot> rootIterator() {
		return _roots.iterator();
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public void validate() {
		try
		{
			for (Iterator<IPackageRoot> iter = rootIterator(); iter.hasNext();) {
				validate(iter.next());
			}
		}
		catch(Exception exception)
		{
			System.out.println(exception.getMessage());
		}
	}

	/**
	 * {@inheritDoc}
	 */
	private void validate(IPackage package$) throws Exception {
		for (Iterator<ICompilationUnit> iter = package$
				.compilationUnitIterator(); iter.hasNext();) {
			iter.next().validate();
		}

		for (Iterator<IPackage> iter = package$.packageIterator(); iter
				.hasNext();) {
			validate(iter.next());
		}
	}
}
